[pt][fbgemm] Turn on USE_FBGEMM on Windows env

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33000 [pt][fbgemm] Turn on USE_FBGEMM on Windows env**

As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

jianyuh committed:
[pt][fbgemm] Turn on USE_FBGEMM on Windows env

As Title says

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

peterjc123 commented:
Wow. The actual error is
```
FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/QuantizedLinear.cpp.obj 
C:\Users\circleci\project\build\win_tmp\bin\sccache.exe  cl   /TP -DAT_PARALLEL_OPENMP=1 -DCPUINFO_SUPPORTED_PLATFORM=1 -DIDEEP_USE_MKL -DMAGMA_V2 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Dtorch_cpu_EXPORTS -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -I..\caffe2\..\torch\csrc\api -I..\caffe2\..\torch\csrc\api\include -I..\caffe2\aten\src\TH -Icaffe2\aten\src\TH -I..\caffe2\..\torch\..\aten\src -Icaffe2\aten\src -Icaffe2\..\aten\src -Icaffe2\..\aten\src\ATen -I..\caffe2\..\torch\csrc -I..\caffe2\..\torch\..\third_party\miniz-2.0.8 -I..\aten\src\TH -I..\aten\..\third_party\catch\single_include -I..\aten\src\ATen\.. -Icaffe2\aten\src\ATen -I..\third_party\miniz-2.0.8 -I..\caffe2\core\nomnigraph\include -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\third_party\cpuinfo\include -I..\third_party\fbgemm\include -I..\third_party\fbgemm -I..\third_party\fbgemm\third_party\asmjit\src -I..\third_party\FP16\include -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I..\third_party\protobuf\src -Iwin_tmp\mkl\include -I..\third_party -I..\cmake\..\third_party\eigen -IC:\Jenkins\Miniconda3\include -IC:\Jenkins\Miniconda3\lib\site-packages\numpy\core\include -I..\cmake\..\third_party\pybind11\include -I\opt\rocm\hip\include -I\include -I..\cmake\..\third_party\cub -Iwin_tmp\magma\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I..\caffe2 /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp -DUSE_FBGEMM -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -DASMJIT_STATIC -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\aten\src\ATen\native\QuantizedLinear.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\aten\src\ATen\native\QuantizedLinear.cpp
..\aten\src\ATen\native\QuantizedLinear.cpp(304): error C3688: invalid literal suffix 'p'; literal operator or literal operator template 'operator ""p' not found
..\aten\src\ATen\native\QuantizedLinear.cpp(304): error C2789: 'significand': an object of const-qualified type must be initialized
..\aten\src\ATen\native\QuantizedLinear.cpp(304): note: see declaration of 'significand'
```

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

peterjc123 commented on cmake/Dependencies.cmake:
```diff
@@ -524,11 +524,8 @@ if(USE_FBGEMM)
     set(USE_FBGEMM OFF)
   endif()
   if(MSVC)
```
Maybe it's sufficient to remove this if condition.

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

peterjc123 commented:
It is not enabled since I can see this in the log.
```
CMake Warning at cmake/Dependencies.cmake:520 (message):
  A compiler with AVX512 support is required for FBGEMM.  Not compiling with
  FBGEMM.  Turn this warning off by USE_FBGEMM=OFF.
```

peterjc123 committed:
debug avx512

peterjc123 committed:
add missing flag for avx512

peterjc123 committed:
Remove flags

peterjc123 commented:
Looks like we will need to switch to VS 2017 14.12 or VS 2019 to enable AVX512.

peterjc123 committed:
Switch to VS 14.12

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh commented:
It appears the compilation errors have been fixed, but now I have the link errors as shown in https://circleci.com/gh/pytorch/pytorch/4426203?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link :
```
fbgemm.lib(FbgemmFP16.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
fbgemm.lib(FbgemmFloat16ConvertAvx2.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
fbgemm.lib(FbgemmFP16UKernelsIntrinsicAvx2.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
fbgemm.lib(FbgemmFP16UKernelsIntrinsicAvx512.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
fbgemm.lib(FbgemmFP16UKernelsIntrinsicAvx512_256.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
fbgemm.lib(Utils.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
fbgemm.lib(FbgemmFloat16Convert.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
fbgemm.lib(UtilsAvx2.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
fbgemm.lib(UtilsAvx512.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
fbgemm.lib(FbgemmFloat16ConvertAvx512.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in CPUGenerator.cpp.obj
msvcprt.lib(MSVCP140.dll) : error LNK2005: "public: struct _Cvtvec __cdecl std::_Locinfo::_Getcvt(void)const " (?_Getcvt@_Locinfo@std@@QEBA?AU_Cvtvec@@XZ) already defined in fbgemm.lib(Utils.cc.obj)
msvcprt.lib(MSVCP140.dll) : error LNK2005: "public: int __cdecl std::basic_streambuf<char,struct std::char_traits<char> >::sputc(char)" (?sputc@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAAHD@Z) already defined in fbgemm.lib(Utils.cc.obj)
msvcprt.lib(MSVCP140.dll) : error LNK2005: "public: __int64 __cdecl std::basic_streambuf<char,struct std::char_traits<char> >::sputn(char const *,__int64)" (?sputn@?$basic_streambuf@DU?$char_traits@D@std@@@std@@QEAA_JPEBD_J@Z) already defined in fbgemm.lib(Utils.cc.obj)
msvcprt.lib(MSVCP140.dll) : error LNK2005: "protected: char * __cdecl std::basic_streambuf<char,struct std::char_traits<char> >::_Pninc(void)" (?_Pninc@?$basic_streambuf@DU?$char_traits@D@std@@@std@@IEAAPEADXZ) already defined in fbgemm.lib(Utils.cc.obj)
msvcprt.lib(MSVCP140.dll) : error LNK2005: "public: void __cdecl std::basic_ios<char,struct std::char_traits<char> >::setstate(int,bool)" (?setstate@?$basic_ios@DU?$char_traits@D@std@@@std@@QEAAXH_N@Z) already defined in fbgemm.lib(Utils.cc.obj)
msvcprt.lib(MSVCP140.dll) : error LNK2005: "public: void __cdecl std::basic_ostream<char,struct std::char_traits<char> >::_Osfx(void)" (?_Osfx@?$basic_ostream@DU?$char_traits@D@std@@@std@@QEAAXXZ) already defined in fbgemm.lib(Utils.cc.obj)
msvcprt.lib(MSVCP140.dll) : error LNK2005: "public: class std::basic_ostream<char,struct std::char_traits<char> > & __cdecl std::basic_ostream<char,struct std::char_traits<char> >::operator<<(int)" (??6?$basic_ostream@DU?$char_traits@D@std@@@std@@QEAAAEAV01@H@Z) already defined in fbgemm.lib(Utils.cc.obj)
...
```

It seems to be related to the mismatch of static/shared build mode of FBGEMM and PyTorch. One was linked against the CRT DLL (/MD) and the other was linked statically (/MT). I am checking the configuration of static/shared library setting of FBGEMM/PyTorch: we use `BUILD_SHARED_LIBS=ON/OFF` to configure this with PyTorch and `FBGEMM_LIBRARY_TYPE=static/shared` with FBGEMM.

Any insights about how to fix this? Thanks! cc @peterjc123 @shz0116

shz0116 commented:
@jianyuh Based on my experience, for static build, should use /MT, dynamic build should use /MD.
It looks you build FBGEMM for static, while CPUGenerator.cpp.obj is built using dynamic. @peterjc123

shz0116 commented:
You can change FBGEMM to shared for a test.

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh commented:
@shz0116 : If I change FBGEMM_LIBRARY_TYPE from "static" to "shared", it shows the following error in the beginning (e.g., https://circleci.com/gh/pytorch/pytorch/4428668?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link):
```
Feb 06 00:34:59 CMake Error: install(EXPORT "Caffe2Targets" ...) includes target "torch_cpu" which requires target "fbgemm" that is not in the export set.
Feb 06 00:35:00 CMake Warning at caffe2/CMakeLists.txt:622 (add_library):
```

> You can change FBGEMM to shared for a test.

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

peterjc123 commented:
> @shz0116 : If I change FBGEMM_LIBRARY_TYPE from "static" to "shared", it shows the following error in the beginning (e.g., https://circleci.com/gh/pytorch/pytorch/4428668?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link):
> 
> ```
> Feb 06 00:34:59 CMake Error: install(EXPORT "Caffe2Targets" ...) includes target "torch_cpu" which requires target "fbgemm" that is not in the export set.
> Feb 06 00:35:00 CMake Warning at caffe2/CMakeLists.txt:622 (add_library):
> ```
> 
> > You can change FBGEMM to shared for a test.

You may need to install the target in shared builds.

jianyuh commented:
@peterjc123 : Do you know how I can change the installation way of the FBGEMM target in shared builds? Any place in ".circleci" folder? Sorry I am not familiar with this... Thanks!

peterjc123 committed:
try fix fbgemm shared build

peterjc123 committed:
Fix typo

peterjc123 commented:
@jianyuh Tried with 4508ff3 and ec3ab57.

peterjc123 committed:
attempt 2

peterjc123 committed:
fix typo

peterjc123 committed:
Fix export on Windows

peterjc123 commented:
@shz0116 @jianyuh I was able to fix the configuration of the CMake scripts by modifying the `CMakeList.txt` at FBGEMM. The changes are listed as the following:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 39d597c..373b5a8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -274,7 +274,7 @@ target_include_directories(fbgemm BEFORE
     PUBLIC $<BUILD_INTERFACE:${FBGEMM_SOURCE_DIR}>
     PUBLIC $<BUILD_INTERFACE:${FBGEMM_SOURCE_DIR}/include>)

-target_link_libraries(fbgemm asmjit cpuinfo)
+target_link_libraries(fbgemm $<BUILD_INTERFACE:asmjit> $<BUILD_INTERFACE:cpuinfo>)
 add_dependencies(fbgemm asmjit cpuinfo)

 install(TARGETS fbgemm EXPORT fbgemmLibraryConfig
@@ -285,6 +285,9 @@ install(TARGETS fbgemm EXPORT fbgemmLibraryConfig
 install(FILES ${FBGEMM_PUBLIC_HEADERS}
   DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/fbgemm")

+install(EXPORT fbgemmLibraryConfig DESTINATION share/cmake/fbgemm
+  FILE fbgemmLibraryConfig.cmake)
+
 #Make project importable from the build directory
 #export(TARGETS fbgemm asmjit FILE fbgemmLibraryConfig.cmake

```

peterjc123 committed:
Remove comments

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh commented:
Add a PR (https://github.com/pytorch/pytorch/pull/33059) for testing the current PyTorch changes + FBGEMM `CMakeList.txt` change.

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh commented:
We have made progress on the linking stage after @peterjc123 's fixes for static/shared library build. Now we have more issues on using `dllimport` and `dllexport` in C++ classes on Windows:
https://circleci.com/gh/pytorch/pytorch/4433071?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh commented:
@peterjc123 : The `dllimport` and `dllexport` are fixed by adding `CAFFE2_API`. I still have some remaining issues on `error LNK2019: unresolved external symbol "void __cdecl ` issue:

```
channel_shuffle_dnnlowp_op.cc.obj : error LNK2019: unresolved external symbol "void __cdecl fbgemm::transpose_4rows(int,unsigned char const *,unsigned char *)" (?transpose_4rows@fbgemm@@YAXHPEBEPEAE@Z) referenced in function "public: bool __cdecl caffe2::ChannelShuffleDNNLowPOp<unsigned char>::RunOnDeviceWithOrderNHWC$omp$1(void)" (?RunOnDeviceWithOrderNHWC$omp$1@?$ChannelShuffleDNNLowPOp@E@caffe2@@QEAA_NXZ)
elementwise_add_dnnlowp_op.cc.obj : error LNK2019: unresolved external symbol "void __cdecl caffe2::internal::ElementWiseSumAVX2<unsigned char,0>(unsigned char const *,unsigned char const *,unsigned char *,int,float,int,float,int,float,int)" (??$ElementWiseSumAVX2@E$0A@@internal@caffe2@@YAXPEBE0PEAEHMHMHMH@Z) referenced in function "public: virtual bool __cdecl caffe2::AddDNNLowPOp<unsigned char>::RunOnDevice$omp$1(void)" (?RunOnDevice$omp$1@?$AddDNNLowPOp@E@caffe2@@UEAA_NXZ)
```

More details can be found in https://circleci.com/gh/pytorch/pytorch/4436134?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link 

It appears to be related to C / C++ compiler issue: https://stackoverflow.com/questions/13099083/visual-c-error-lnk2019

Any insights about fixing this? Do I have to add something like
```
#ifdef __cplusplus
extern "C" {
#endif
```
in those *.cc files?

Thanks!

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

peterjc123 commented:
> @peterjc123 : The `dllimport` and `dllexport` are fixed by adding `CAFFE2_API`. I still have some remaining issues on `error LNK2019: unresolved external symbol "void __cdecl ` issue:
> 
> ```
> channel_shuffle_dnnlowp_op.cc.obj : error LNK2019: unresolved external symbol "void __cdecl fbgemm::transpose_4rows(int,unsigned char const *,unsigned char *)" (?transpose_4rows@fbgemm@@YAXHPEBEPEAE@Z) referenced in function "public: bool __cdecl caffe2::ChannelShuffleDNNLowPOp<unsigned char>::RunOnDeviceWithOrderNHWC$omp$1(void)" (?RunOnDeviceWithOrderNHWC$omp$1@?$ChannelShuffleDNNLowPOp@E@caffe2@@QEAA_NXZ)
> elementwise_add_dnnlowp_op.cc.obj : error LNK2019: unresolved external symbol "void __cdecl caffe2::internal::ElementWiseSumAVX2<unsigned char,0>(unsigned char const *,unsigned char const *,unsigned char *,int,float,int,float,int,float,int)" (??$ElementWiseSumAVX2@E$0A@@internal@caffe2@@YAXPEBE0PEAEHMHMHMH@Z) referenced in function "public: virtual bool __cdecl caffe2::AddDNNLowPOp<unsigned char>::RunOnDevice$omp$1(void)" (?RunOnDevice$omp$1@?$AddDNNLowPOp@E@caffe2@@UEAA_NXZ)
> ```
> 
> More details can be found in https://circleci.com/gh/pytorch/pytorch/4436134?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
> 
> It appears to be related to C / C++ compiler issue: https://stackoverflow.com/questions/13099083/visual-c-error-lnk2019
> 
> Any insights about fixing this? Do I have to add something like
> 
> ```
> #ifdef __cplusplus
> extern "C" {
> #endif
> ```
> 
> in those *.cc files?
> 
> Thanks!

Looks like we will need to add more `CAFFE2_API`s, like `ElementWiseSumAVX2` in `caffe2\quantization\server\elementwise_sum_dnnlowp_op_avx2.cc`.

jianyuh committed:
Update on "[pt][fbgemm] Turn on USE_FBGEMM on Windows env"


As Title says. FBGEMM has recently added the support for Windows.

Differential Revision: [D19738268](https://our.internmc.facebook.com/intern/diff/D19738268/)

[ghstack-poisoned]

jianyuh commented:
This PR is switched to https://github.com/pytorch/pytorch/pull/33250
